src/control/Control.ts

```diff
@@ -1,88 +1,106 @@
 
   public init(flicking: Flicking): this {
     this._flicking = flicking;
-    this._controller = new AxesController(flicking);
+    this._controller.init(flicking);
 
     return this;
   }
 
   public destroy(): this {
-    this._controller?.destroy();
+    this._controller.destroy();
 
     this._flicking = null;
-    this._controller = null;
-    this._activeIndex = 0;
+    this._activePanel = null;
 
     return this;
   }
 
-  public getController() { return this._controller; }
-  public getActiveIndex() { return this._activeIndex; }
-  public isAnimating() { return this._controller?.getState().playing || false; }
+  public get controller() { return this._controller; }
+  public get activeIndex() { return this._activePanel?.index ?? -1; }
+  public get activePanel() { return this._activePanel; }
+  public get animating() { return this._controller.state.playing; }
+  public get holding() { return this._controller.state.holding; }
 
   public enable(): this {
-    this._controller?.enable();
+    this._controller.enable();
 
     return this;
   }
 
   public disable(): this {
-    this._controller?.disable();
+    this._controller.disable();
 
     return this;
   }
 
-  @requireFlicking("Control")
-  public updateInput() {
-    this._controller!.update();
+  public updateInput(): this {
+    this._controller.update();
+
+    return this;
   }
 
-  @requireFlicking("Control")
-  public async moveToPanel(panel: Panel, duration: number) {
-    const flicking = this._flicking!;
+  public checkActivePanelIsRemoved(): this {
+    if (this._activePanel?.isRemoved()) {
+      this._activePanel = null;
+    }
 
-    const camera = flicking.getCamera();
+    return this;
+  }
+
+  public async moveToPanel(panel: Panel, duration: number, axesEvent?: OnRelease) {
+    const flicking = getFlickingAttached(this._flicking, "Control");
+    const camera = flicking.camera;
 
-    const triggeringEvent = panel.getIndex() !== this._activeIndex ? EVENTS.CHANGE : EVENTS.RESTORE;
+    const triggeringEvent = panel !== this._activePanel ? EVENTS.CHANGE : EVENTS.RESTORE;
 
     const eventSuccess = flicking.trigger(triggeringEvent, {
-      index: panel.getIndex(),
+      index: panel.index,
       panel,
-      isTrusted: false,
-      direction: panel.getPosition() > camera.getPosition() ? DIRECTION.NEXT : DIRECTION.PREV
+      isTrusted: axesEvent?.isTrusted || false,
+      direction: getDirection(camera.position, panel.position)
     });
 
     if (!eventSuccess) {
-      return Promise.resolve();
+      // TODO: Add new error type that shows it rejected by user stop()
+      return Promise.reject();
     }
 
-    return this._controller!.animateTo(panel.getPosition(), duration)
-      .then(() => {
-        this._activeIndex = panel.getIndex();
-      });
+    const updateActivePanel = () => {
+      this._activePanel = panel;
+    };
+    const animate = () => this._controller.animateTo(panel.position, duration, axesEvent);
+
+    if (duration === 0) {
```

Comment:
When duration is 0, it seems to be a problem if the order of operation of animate => updateActivePanel is changed.

Since the duration is 0, there seems to be no problem right now, but if there is activePanel related code in animateTo, it will be a problem.

Reply:
That's an exceptional case.
As it's promise based method, it guarantees that it returns promise while updating the index synchronously.
And there won't be code that manipulates the active panel in AxesController, that's definitely not a role of it.

Reply:
You mean you have to compare if duration === 0?

I thought it was the same code because both return promises.

Reply:
That code is to match Axes's behavior.
Axes runs synchronously if the duration is 0, so it just updates the active index synchronously before the Promise is resolved.
It makes sure that the index is updated immediately after calling moveTo with duration 0.

```ts
if (duration > 0) {
  this.animateTo(movedPos, duration);
} else {
  this.em.triggerChange(movedPos);
  this.finish(false);
}
```
> Axes setTo